core/services/pipeline/orm.go

```diff
@@ -0,0 +1,360 @@
+                LEFT JOIN pipeline_task_specs AS specs_left ON specs_left.id = specs_right.successor_id
+                LEFT JOIN pipeline_task_runs AS successors ON successors.pipeline_task_spec_id = specs_left.id
+                      AND successors.pipeline_run_id = pipeline_task_runs.pipeline_run_id
+                WHERE successors.id = ?`, ptRun.ID).Find(&predecessors).Error
+		if err != nil {
+			return errors.Wrap(err, "error finding task run predecessors")
+		}
+
+		// Get the job ID
+		var job struct{ ID int32 }
+		err = tx.Raw(`
+            SELECT jobs.id FROM pipeline_task_runs
+            LEFT JOIN pipeline_task_specs ON pipeline_task_specs.id = pipeline_task_runs.pipeline_task_spec_id
+            LEFT JOIN jobs ON jobs.pipeline_spec_id = pipeline_task_specs.pipeline_spec_id
+            WHERE pipeline_task_runs.id = ?
+        `, ptRun.ID).Scan(&job).Error
+		if err != nil {
+			return errors.Wrap(err, "error finding job ID")
+		}
+
+		// Call the callback
+		result := fn(job.ID, ptRun, predecessors)
+
+		// Update the task run record with the output and error
+		if result.Value != nil {
+			out := &JSONSerializable{Val: result.Value}
+			err = tx.Exec(`UPDATE pipeline_task_runs SET output = ?, error = NULL, finished_at = ? WHERE id = ?`, out, time.Now(), ptRun.ID).Error
+		} else if result.Error != nil {
+			logger.Errorw("Error in pipeline task", "error", result.Error)
+			errString := null.StringFrom(result.Error.Error())
+			err = tx.Exec(`UPDATE pipeline_task_runs SET output = NULL, error = ?, finished_at = ? WHERE id = ?`, errString, time.Now(), ptRun.ID).Error
+		}
+		if err != nil {
+			return errors.Wrap(err, "could not mark pipeline_task_run as finished")
+		}
+		return nil
+	})
+	return done, err
+}
+
+const (
+	postgresChannelRunStarted   = "pipeline_run_started"
+	postgresChannelRunCompleted = "pipeline_run_completed"
+)
+
+func (o *orm) ListenForNewRuns() (*utils.PostgresEventListener, error) {
+	listener := &utils.PostgresEventListener{
+		URI:                  o.uri,
+		Event:                postgresChannelRunStarted,
+		MinReconnectInterval: 1 * time.Second,
+		MaxReconnectDuration: 1 * time.Minute,
+	}
+	err := listener.Start()
+	if err != nil {
+		return nil, err
+	}
+	return listener, nil
+}
+
+// AwaitRun waits until a run has completed (either successfully or with errors)
+// and then returns.  It uses two distinct methods to determine when a job run
+// has completed:
+//    1) periodic polling
+//    2) Postgres notifications
+func (o *orm) AwaitRun(ctx context.Context, runID int64) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// This goroutine polls the DB at a set interval
+	chPoll := make(chan error)
+	go func() {
+		var err error
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			var done bool
+			done, err = runFinished(o.db, runID)
+			if err != nil || done {
+				break
+			}
+			time.Sleep(1 * time.Second)
```

Comment:
Perhaps a timer channel alongside the ctx.Done() might be nicer?

And move the poll interval to a constant?

Check pattern in EthBroadcaster

Reply:
This has come up enough recently that I think I might make a `utils.TimerSleeperTask`